skbeam/core/correlation.py

```diff
@@ -1027,55 +1062,74 @@ def __call__(self, img1, img2=None, normalization=None):
-                                    ] = img1.ravel()[self.pxlsts[i]]
+            index_start, index_stop = idpos[i], idpos[i+1]
+            ppiis = self.ppii[index_start:index_stop]
+            ppjjs = self.ppjj[index_start:index_stop]
+            pis = self.pi[index_start:index_stop]
+            pjs = self.pj[index_start:index_stop]
+            tmpimg = np.zeros(self.shapes[i, :])
+            tmpimg[ppiis, ppjjs] = img1[pis, pjs]
+
             if not self_correlation:
-                self.tmpimgs2[i] *= 0
-                self.tmpimgs2[i].ravel()[
-                                        self.subpxlsts[i]
-                                        ] = img2.ravel()[self.pxlsts[i]]
+                tmpimg2 = np.zeros_like(tmpimg)
+                tmpimg2[ppiis, ppjjs] = img2[pis, pjs]
+                tmpimg2[ppiis, ppjjs] = img2[pis, pjs]
 
-            # multiply by maskcorrs > 0 to ignore invalid regions
             if self_correlation:
-                ccorr = _cross_corr(self.tmpimgs[i])*(self.maskcorrs[i] > 0)
+                ccorr = _cross_corr(tmpimg)
             else:
-                ccorr = _cross_corr(self.tmpimgs[i], self.tmpimgs2[i]) * \
-                        (self.maskcorrs[i] > 0)
+                ccorr = _cross_corr(tmpimg, tmpimg2)
 
-            # now handle the normalizations
+            # Note, in this code, non-overlapping regions will now get np.nan
+            # also, for sym averaging, if Icorr*Icorr2==0, then we also get
+            # np.nan
             if 'symavg' in normalization:
                 # do symmetric averaging
-                Icorr = _cross_corr(self.tmpimgs[i] *
-                                    self.submasks[i], self.submasks[i])
+                Icorr = _cross_corr(tmpimg * self.submasks[i],
+                                    self.submasks[i])
                 if self_correlation:
-                    Icorr2 = _cross_corr(self.submasks[i], self.tmpimgs[i] *
+                    Icorr2 = _cross_corr(self.submasks[i], tmpimg *
                                          self.submasks[i])
                 else:
-                    Icorr2 = _cross_corr(self.submasks[i], self.tmpimgs2[i] *
+                    Icorr2 = _cross_corr(self.submasks[i], tmpimg2 *
                                          self.submasks[i])
-                # there is an extra condition that Icorr*Icorr2 != 0
-                w = np.where(np.abs(Icorr*Icorr2) > 0)
-                ccorr[w] *= self.maskcorrs[i][w]/Icorr[w]/Icorr2[w]
+                ccorr *= self.maskcorrs[i]/Icorr/Icorr2
 
             if 'regular' in normalization:
-                # only run on overlapping regions for correlation
-                w = self.pxlst_maskcorrs[i]
-                ccorr[w] /= self.maskcorrs[i][w] * \
-                    np.average(self.tmpimgs[i].
-                               ravel()[self.subpxlsts[i]])**2
-
+                if self_correlation:
+                    ccorr /= self.maskcorrs[i] * \
```

Comment:
the code has been changed to that non-overlapping regions are `np.nan`, for the record